mfr/extensions/tabular/libs/stdlib_tools.py

```diff
@@ -92,3 +115,62 @@ def _set_dialect_quote_attrs(dialect, data):
             dialect.quotechar = '"'
         if re.search('"""[[({]\'.+\',', data):
             dialect.doublequote = True
+
+
+def _trim_or_append_data(fp, text, read_size, sniff_size):
+    """Recursively read data from a file and return its first row. The file starts with ``text``
+    and the file pointer points to the next character immediately after `text`.
+
+    :param fp: the file pointer from which data is read
+    :param text: the current text chunk to check the new line character
+    :param read_size: the last read size when `fp.read()` is called
+    :param sniff_size: the accumulated size fo the text to sniff
+    :return: the first row of the file in string
+    """
+
+    logger.info('>>> ??? &&& ~~~ _trim_or_append_data() ...')
+    logger.info('>>> ??? &&& ~~~ len(text)={}\tread_size={}\tsniff_size={}'
+                .format(len(text), read_size, sniff_size))
+
+    # Try to find the first new line character in the text chunk
+    index = _find_new_line(text)
+    # If found, return the trimmed substring
+    if index != -1:
+        logger.info('>>> ??? &&& ~~~ new line found @ index = {}, '
+                    'return the trimmed text'.format(index))
+        return text[:index]
+    # Otherwise, update `sniff_size` and then sniff more (2 times of the last `read_size`) text
+    sniff_size += read_size
+    read_size *= 2
+    more_text = fp.read(read_size)
+
+    # If text to sniff now goes over the max file size limit, raise the renderer error since there
+    # is no need to sniff when the file is already too large to be rendered.
+    if sniff_size + len(more_text) >= MAX_FILE_SIZE:
```

Comment:
This is a redundant check since the renderer check file size in advance before handing off to the sniffer. However, I prefer to keep it since it is logically sound and necessary if we look at the recursion by itself.